language/stdlib/modules/ValidatorConfig.move

```diff
@@ -149,5 +149,102 @@ module ValidatorConfig {
     public fun get_validator_network_address(config_ref: &Config): &vector<u8> {
         &config_ref.validator_network_address
     }
+
+    // **************** SPECIFICATIONS ****************
+
+    /// # Module specifications
+
+    spec module {
+        pragma verify = true, aborts_if_is_strict = true;
+
+        /// Returns true if a ValidatorConfig resource exists under addr.
+        define spec_exists_config(addr: address): bool {
+            exists<ValidatorConfig>(addr)
+        }
+
+        /// Returns true if addr has an operator account.
+        define spec_has_operator(addr: address): bool {
+            Option::spec_is_some(global<ValidatorConfig>(addr).operator_account)
+        }
+
+        /// Returns the operator accounr of a validator if it has one,
+        /// and returns the addr itself otherwise.
+        define spec_get_operator(addr: address): address {
+            if (spec_has_operator(addr)) {
+                Option::spec_value_inside(global<ValidatorConfig>(addr).operator_account)
+            } else {
+                addr
+            }
+        }
+
+        /// Returns true if there a config published under addr.
+        define spec_has_config(addr: address): bool {
+            Option::spec_is_some(global<ValidatorConfig>(addr).config)
+        }
+
+        /// Returns the config published under addr.
+        define spec_get_config(addr: address): Config {
+            Option::spec_value_inside(global<ValidatorConfig>(addr).config)
+        }
+
+        /// Returns true if addr is a valid validator.
+        define spec_is_valid(addr: address): bool {
+            spec_exists_config(addr) && spec_has_config(addr)
+        }
+    }
+
+    /// ## Validator stays valid once it becomes valid
+
+    spec schema ValidatorStaysValid {
+        ensures forall validator: address:
+            old(spec_is_valid(validator)) ==> spec_is_valid(validator);
+    }
+
+    spec module {
+        apply ValidatorStaysValid to *;
+    }
+
+    /// # Specifications for individual functions
+
+    spec fun publish {
```

Comment:
I would suggest to move the function specs right after the function. As well as any helper functions related to a given function `f` (like `define spec_f`). This makes it so much easier to maintain the code. It's already done in various places.

If you do so, you should use the following ordering:

```
fun foo(...): ... {
}
spec fun foo {
   // Having this first will give doc generator the context where to associate the following specs with
   ...
}
spec schema FooS {
   // Any schemas belonging to this function
}
...
spec module {
   // Any spec functions related to this function
   define spec_foo(...): ... { ... }
}
...
```

```